Remove TransactionDetails from Wallet API

notmandatory opened the pull request:
### Description

Removed `TransactionDetails` and changed `Wallet::get_tx` to return a `CanonicalTx`, and `TxBuilder::finish` to return only a `PartiallySignedTransaction`. This should fix #922 and fix #1015.  

I also added `Wallet` functions to get a `Transaction` send and receive amounts, fee, and `FeeRate`.

see: https://github.com/bitcoindevkit/bdk/issues/922#issuecomment-1652814975

### Notes to the reviewers

Alot of wallet tests had to change since `TxBuilder::finish` only returns a PSBT now. 

I added a new `CalculateFeeError` which follows changes coming in #1028.

### Changelog notice

Added
- Wallet::sent_and_received function
- Wallet::calculate_fee and Wallet::calculate_fee_rate functions
- Wallet::error::CalculateFeeError
- Wallet::insert_txout function to allow inserting foreign TxOuts

BREAKING CHANGES:

Removed
- TransactionDetails struct

Changed
- Wallet::get_tx now returns CanonicalTx instead of TransactionDetails
- TxBuilder::finish now returns only a PartiallySignedTransaction

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

notmandatory committed:
feat(wallet)!: remove TransactionDetails from bdk::Wallet API

Added
- Wallet::sent_and_received function
- Wallet::calculate_fee and Wallet::calculate_fee_rate functions
- Wallet::error::CalculateFeeError

BREAKING CHANGES:

Removed
- TransactionDetails struct

Changed
- Wallet::get_tx now returns CanonicalTx instead of TransactionDetails
- TxBuilder::finish now returns only a PartiallySignedTransaction

LLFourn commented:
I think the plan is to deprecate `TransactionDetails` altogether. What about just doing #1015 or is the plan to change this test to test the things in #1015 when they're added?

EDIT: this comment was made before seeing the discussion in #1015 -- I've commented over there.

notmandatory commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -427,27 +428,97 @@ impl<D> Wallet<D> {
+
+    /// Return `TransactionAmounts` for a `Transaction` in relation to the `Wallet` and it's
+    /// descriptors.
+    pub fn sent_and_received(&self, tx: &Transaction) -> TransactionAmounts {
```
Can I rename this to `tx_amounts`? the name `sent_and_received` is a bit long an awkward.

danielabrozzoni replied:
ACK

evanlinjin replied:
Personally I prefer `sent_and_received` as I find it more descriptive.

danielabrozzoni commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -427,27 +427,97 @@ impl<D> Wallet<D> {
+                    None
+                }
+            })
+            .next()
```
nit: you can use `find_map`
```
    pub fn get_txout(&self, outpoint: &OutPoint) -> Option<TxOut> {
        self.indexed_graph
            .graph()
            .all_txouts()
            .find_map(|(op, txo)| {
                if op.eq(outpoint) {
                    Some(txo.clone())
                } else {
                    None
                }
            })
    }
```

notmandatory replied:
Thanks for the tip! but while fixing it I found that there's already a `TxGraph::get_txout()` function so I'm using that instead and was able to remove this whole function.

danielabrozzoni reviewed (approved):
I just have a nit 

ACK 40d5af39e256c2df3ea468e5f6cc32e5898cc77d

notmandatory commented:
@LLFourn continuing https://github.com/bitcoindevkit/bdk/issues/1015#issuecomment-1657935215 here. 

Are you ok with the name `TotalSentReceived` for the struct and `total_sent_received` for the function? I just don't think `and` in the names is as useful as `total`. But if you really hate it I can live with `SentAndReceived`/`sent_and_received`.  

I'm now calling the `TxGraph::calculate_fee` function but added a friendlier `Result` return type and new `CalculateFeeError`. If you like this approach should I move it down to TxGraph?

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -427,27 +427,49 @@ impl<D> Wallet<D> {
+        match self.indexed_graph.graph().calculate_fee(tx) {
+            None => Err(CalculateFeeError::MissingTxOut),
+            Some(fee) if fee < 0 => Err(CalculateFeeError::NegativeFee(fee)),
+            Some(fee) => Ok(u64::try_from(fee).unwrap()),
```
I think this API is better and it should just be pushed down to `TxGrpah`. @evanlinjin?

evanlinjin replied:
I think that will be nice. To improve the API further, I would have the `CalculateFeeError::MissingTxOut` specify the missing txout outpoints as well.

LLFourn commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -334,4 +343,33 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
         input_matches || output_matches
     }
+
+    /// Return `TotalSentReceived` for a `Transaction` in relation to the `SpkTxOutIndex`.
```
Isn't this meant to replace `sent_and_received`? (it's still here).
After seeing this I think the tuple `(u64, u64)` as it was before since  has the order of the returned value is in the name `total_sent_received` which implies `sent` is being returned before received. Not a major concern so feel free to go with whatever feels right.

In any case please just rename the existing function `sent_and_received` and change the output type (if you still want a custom type).

notmandatory replied:
Hey sorry I just realized you already had the `sent_and_received` function implemented in the `chain` crate. Makes sense so I just exposed it as it is there in the `bdk` API with the same name and tuple result type.

LLFourn reviewed (commented):
Concept ACK. See minor comments.

notmandatory committed:
feat(tx_graph)!: change TxGraph::calculate_fee to return Result<u64,CalculateFeeError>

added
- tx_graph::CalculateFeeError enum

BREAKING CHANGES:

changed
- TxGraph::calculate_fee function to return Result<u64,CalculateFeeError> instead of Option<i64>

LLFourn commented on crates/chain/tests/test_tx_graph.rs:
```diff
@@ -474,7 +482,7 @@ fn test_calculate_fee_on_coinbase() {
     let graph = TxGraph::<()>::default();
 
-    assert_eq!(graph.calculate_fee(&tx), Some(0));
+    assert_matches!(graph.calculate_fee(&tx), Ok(0));
```
nit `assert_eq` here? I think you could make all of these `assert_eq` and avoid the dev dependency.

notmandatory replied:
Fixed, I was getting an error before but it was because I forgot to derive `Eq`, `PartialEq` for `CalculateFeeError`, it works now without the extra dev dep.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -969,20 +991,8 @@ impl<D> Wallet<D> {
-            confirmation_time: ConfirmationTime::Unconfirmed { last_seen: 0 },
-            received,
-            sent,
-            fee: Some(fee_amount),
```
I am concerned that we can't reproduce this `fee` value unless we return it here. When you use `add_foreign_utxo` to the builder you are telling it the value of the input. This won't exist in the `TxGraph` (although it could we have a `mut self` here!).

So we can:
1) add back the second return value to return the fee
2) Keep only one return value but insert the `TxOut`s into the `TxGraph` so we can later query for fee and get the right answer (they would be un-anchored  i.e. assumed to be in the mempool). Thse txouts would be committed to disk so we could query the fee of this tx later on.
3) Require the user inserts the foreign utxo into the wallet manually before they can use it in `add_foreign_utxo`.
4) Simply recommend that the user always insert the foreign UTXO into the wallet manually if they want to be able to calculate the fees of the resulting transaction. This makes the user think about whether they can actually trust the values they have in this UTXO.

I think I lean towards (4) because we have planning module integration coming up and that may change transaction building a lot. I notice that in order to do (3) or (4) we'd need to add `insert_txout` method on wallet that calls the graph's method.

notmandatory replied:
Interesting problem. I agree the last option (4) makes sense, I can add some documentation to explain the situation. I'll also add a test to reproduce and confirm it works as expected.

evanlinjin replied:
The other option will change the `TxGraph::calculate_fee` method to include missing outputs.

```rust
pub fn calculate_fee(
    &self, 
    tx: &Transactions, 
    missing_outputs: &HashMap<OutPoint, TxOut>,
) -> Result<i64, CalculateFeeError> {
    unimplemented!()
}
```

Edit: rethinking about it, I think 4 is still the most practical solution. My suggestion will by clunky after a tx is created and the caller wants to calculate the fee.

notmandatory replied:
In a future `1.x` release we could add a `calculate_fee_with_outputs` function is there's any demand for it.

notmandatory replied:
To test the wallet calculate_fee  and fee_rate with and without foreign utxos I added `Wallet::insert_txout`, which exposes the underlying TxGraph::insert_txout. For this new function I'm also persisting the change, as long as the txout data is trusted this is OK isn't it? 

I also updated docs with warnings about missing utxos, and to trust any added. Is it clear enough?

notmandatory commented:
Rebased post #1023 and #1034 being merged.

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -428,27 +429,48 @@ impl<D> Wallet<D> {
-    /// Optionally fill the [`TransactionDetails::transaction`] field with the raw transaction if
-    /// `include_raw` is `true`.
-    pub fn get_tx(&self, txid: Txid, include_raw: bool) -> Option<TransactionDetails> {
+    /// Note `tx` does not have to be in the graph for this to work.
```
Should we note that we need the outputs that this tx spends to exist in the internal tx graph for this to work?

notmandatory replied:
Yup, I'll add the same note I'm adding to the `TxGraph::calculate_fee` to any functions that use it.

LLFourn replied:
Nit: I'm concerned that the notion of a "foreign" txout is ill-defined and will be short lived. It should just say here that the txout must have been inserted into the wallet which won't happen naturally if the transaction has inputs that were not owned by this wallet (as you mention you can remedy this by `insert_txout`). The specific documentation that this problem can occur for you when you use `add_foreign_utxo` in the builder should reside on `add_foreign_utxo` to avoid using the term "foreign" here. Perhaps reference `add_foreign_utxo` here and the likely source of your problems in this doc.

notmandatory replied:
How do these changes look? https://github.com/bitcoindevkit/bdk/pull/1048/commits/ec707899049274fe091ae568a068cd6a1ca02dc3

notmandatory replied:
Fixed docs in 465d53cc88b32bf8098dc61e9afdbb28d22b213d (rebased with other suggestions from @evanlinjin ).

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -428,27 +429,48 @@ impl<D> Wallet<D> {
+    }
+
+    /// Return a single `CanonicalTx` made and received by the wallet or `None` if it doesn't
+    /// exist in the wallet
```
Because of the complexity of the `CanonicalTx` structure, I think this method deserves documentation with a bit more detail. Let me know if this will suffice.

```suggestion
    /// Get a single transaction from the wallet as a [`CanonicalTx`] (if the transaction exists).
    ///
    /// `CanonicalTx` contains the full transaction alongside metadata such as:
    /// * Blocks that the transaction is [`Anchor`]ed in. These may or may not be blocks that exist
    ///   in the best chain.
    /// * The [`ChainPosition`] of the transaction in the best chain - whether the transaction is
    ///   confirmed or unconfirmed. If the transaction is confirmed, the anchor which proves the
    ///   confirmation is provided. If the transaction is unconfirmed, the unix timestamp of when
    ///   the transaction was last seen in the mempool is provided.
    ///
    /// ```rust, no_run
    /// use bdk::{chain::ChainPosition, Wallet};
    /// # let wallet: Wallet<()> = todo!();
    /// # let my_txid: bdk::bitcoin::Txid = todo!();
    ///
    /// let canonical_tx = wallet.get_tx(my_txid).expect("panic if tx does not exist");
    ///
    /// // get reference to full transaction
    /// println!("my tx: {:#?}", canonical_tx.node.tx);
    ///
    /// // list all transaction anchors
    /// for anchor in canonical_tx.mode.anchors {
    ///     println!("tx is anchored by block of hash {}", anchor.anchor_block().hash);
    /// }
    ///
    /// // get confirmation status of transaction
    /// match canonical_tx.observed_as {
    ///     ChainPosition::Confirmed(anchor) => println!(
    ///         "tx is confirmed at height {}, we know this since {}:{} is in the best chain",
    ///         anchor.confirmation_height,
    ///         anchor.anchor_block.height,
    ///         anchor.anchor_block.hash,
    ///     ),
    ///     ChainPosition::Unconfirmed(last_seen) => println!(
    ///         "tx is last seen at {}, it is unconfirmed as it is not anchored in the best chain",
    ///         last_seen,
    ///     ),
    /// }
    /// ```
    ///
    /// [`Anchor`]: bdk_chain::Anchor
```

danielabrozzoni replied:
```
    /// match canonical_tx.observed_as {
    ///     ChainPosition::Confirmed(anchor) => println!(
    ///         "tx is confirmed at height {}, we know this since {}:{} is in the best chain",
    ///         anchor.confirmation_height,
    ///         anchor.anchor_block.height,
    ///         anchor.anchor_block.hash,
    ///     ),
```

Is this correct? I'm still confused about anchors etc, but from my understanding not all the anchors of a transaction have to be in the best chain. A transaction could have an anchor A1 in the best chain, and an anchor A2 which is not in the best chain - in this case, we know that the tx is confirmed because of A1, but we can't say "tx is confirmed at height h2, we know this since A2 is in the best chain"

notmandatory replied:
What I understand this is saying is that if the `canonical_tx.observed_as` is `ChainPosition::Confirmed` then it will contain the anchor that is in the best chain even if there may be other anchors that aren't in the best chain.  If the `canonical_tx` isn't anchored to the best chain it will have `ChainPosition::Unconfirmed` with the last seen time.  The magic looks like it happens in `TxGraph::try_get_chain_position`.

evanlinjin replied:
Yes, the `ChainPosition::Confirmed` variant only contains one anchor, which is the anchor that anchors the tx in the best chain.

notmandatory committed:
feat(wallet): add Wallet::insert_txout function and updated docs for fee functions

added
    - Wallet::insert_txout function to allow inserting foreign TxOuts
    - test to verify error when trying to calculate fee with missing foreign utxo
    - test to calculate fee with inserted foreign utxo

updated
    - docs for Wallet::calculate_fee, Wallet::calculate_fee_rate, and TxGraph::calculate_fee
      with note about missing foreign utxos

evanlinjin commented on crates/chain/src/tx_graph.rs:
```diff
@@ -241,33 +250,50 @@ impl<A> TxGraph<A> {
+            Ok(inputs_sum.0 as i64)
+        } else {
+            Err(CalculateFeeError::MissingTxOut(inputs_sum.1))
+        }?;
```
Nit:
```suggestion
        let (inputs_sum, missing_outputs) = tx.input.iter().fold(
            (0_i64, Vec::new()),
            |(mut sum, mut missing_outpoints), txin| match self.get_txout(txin.previous_output) {
                None => {
                    missing_outpoints.push(txin.previous_output);
                    (sum, missing_outpoints)
                }
                Some(txout) => {
                    sum += txout.value as i64;
                    (sum, missing_outpoints)
                }
            },
        );
        if !missing_outputs.is_empty() {
            return Err(CalculateFeeError::MissingTxOut(missing_outputs));
        }
```

evanlinjin commented on crates/bdk/tests/wallet.rs:
```diff
@@ -88,6 +87,62 @@ fn test_get_funded_wallet_balance() {
+    tx_amounts.sort_by(|a1, a2| a1.0.cmp(&a2.0));
+
+    assert_eq!(tx_amounts.len(), 2);
+    assert_matches!(tx_amounts.get(0), Some((_, (76_000, 50_000))))
```
How did we know what values we are expecting here?

notmandatory replied:
Thanks for catching this, I got sloppy here. Will fix it and add better docs too.

evanlinjin commented on crates/bdk/tests/wallet.rs:
```diff
@@ -88,6 +87,62 @@ fn test_get_funded_wallet_balance() {
+        tx_fee_amounts.get(1),
+        Some((_, Err(CalculateFeeError::MissingTxOut(_))))
+    );
+    assert_matches!(tx_fee_amounts.get(0), Some((_, Ok(1000))))
```
How did we know this fees here? Would it be possible to send a tx from `bitcoind` and have `bitcoind` give us the fee/feerate and we can compare the result with what we get from BDK?

notmandatory replied:
Since we're just calling rust-bitcoin's calculations here I don't think we need to re-validate their code. But I can document the calculation for anyone who wants to see where the final number comes from.

evanlinjin reviewed (changes requested):
I'm liking how this looks. I've got some suggestions to documentation and tests. I'll like to give an ACK after all is addressed (or dismissed).

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -428,27 +429,115 @@ impl<D> Wallet<D> {
+    {
+        let additions = self.indexed_graph.insert_txout(outpoint, &txout);
+        self.persist.stage(ChangeSet::from(additions));
+        self.persist.commit()?;
```
We are not meant to commit here. What if you want to insert a bunch of txouts? This would do IO for every single call.

notmandatory replied:
Makes sense. I removed the commit() and Result<>, and updated the docs.

LLFourn commented on crates/bdk/tests/wallet.rs:
```diff
@@ -1002,11 +1049,13 @@ fn test_add_foreign_utxo() {
         .unwrap();
-    let (mut psbt, details) = builder.finish().unwrap();
+    let mut psbt = builder.finish().unwrap();
+    let fee = psbt.fee_amount();
```
:bulb: : We are using `psbt.fee_amount()` a lot in all of these tests but we also introduced our own fee calc method in this PR. Maybe we can test for free our one against the PSBT one with some macro like `let fee = check_fee!(wallet, psbt)` which would internally assert that psbt.fee_amount was the same as what the wallet calculates and then returns it.

notmandatory replied:
Good idea, I added it in https://github.com/bitcoindevkit/bdk/pull/1048/commits/f1d868d94ae02e6b30771ace01c84b3bed98c839, and was able to remove the test I added for checking fees with a foreign utxo that's now covered by the existing txbuilder foreign utxo tests.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -430,27 +431,113 @@ impl<D> Wallet<D> {
+
+    /// Computes total input value going from script pubkeys in the index (sent) and the total output
+    /// value going to script pubkeys in the index (received) in `tx`. For the `sent` to be computed
+    /// correctly, the output being spent must have already been scanned by the index. Calculating
```
```suggestion
    /// correctly, the outputs being spent must have already been scanned by the index. Calculating
```

LLFourn reviewed (approved):
ACK 1992c62efd4c68a062f91f9d02da685c25ba1b40

left some nits

notmandatory committed:
docs(wallet): update docs for calculate_fee/fee_rate and add_foreign_utxo

notmandatory committed:
test(wallet): add check_fee!(wallet,psbt) macro and use it in place of psbt.fee_amount()

- removed test_calculate_fee_with_inserted_foreign_utxo() since it now duplicates test in test_add_foreign_utxo()

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -430,27 +431,118 @@ impl<D> Wallet<D> {
+    /// value going to script pubkeys in the index (received) in `tx`. For the `sent` to be computed
+    /// correctly, the outputs being spent must have already been scanned by the index. Calculating
+    /// received just uses the [`Transaction`] outputs directly, so it will be correct even if it has
+    /// not been scanned.
```
Nit: I think it will be better to have two sections for these docs.

https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components

```suggestion
    /// Computes total input value going from script pubkeys in the index (sent) and the total output
    /// value going to script pubkeys in the index (received) in `tx`.
    ///
    /// For the `sent` to be computed correctly, the outputs being spent must have already been scanned
    /// by the index. Calculating received just uses the [`Transaction`] outputs directly, so it will
    /// be correct even if it has not been scanned.
```

notmandatory replied:
I hadn't read the documentation guide, good suggestion. I also added some simple examples.

notmandatory committed:
ci: fix msrv dependency versions for rustls

notmandatory commented:
I've rebased, fixed docs and fixed a new MSRV issue.  Just need @evanlinjin to ACK and then I'll merge, which will also fix `master` branch CI.

evanlinjin reviewed (approved):
ACK 5fb5061645ae92d37091a215e48e84423fea48a3
